Add option for greek

Katerina-codes opened the pull request:
New feature to give the user the opportunity to play the game in Greek.

- Need to add README.
- Need to fix bug that adds on 'wins' when it's a draw it Greek.
- Need to refactor to remove some if statements which make method confusing to read.
- Need to extract conversion logic to Conversion class. 

Fixes #7 

@MollieS @jsuchy @felipesere @Maikon

Katerina-codes committed:
Extracted hardcoded strings into methods

- Created Language interface.
- Created English class.

Katerina-codes committed:
Removed hardcoded Moves array from ComputerPlayer

- Moved random move generation logic from ComputerPlayer to Moves enum.

Katerina-codes committed:
Added EnglishTest class and added tests

Katerina-codes committed:
Added Greek class and GreekTest class

- Method prompts for player one in Greek.

Katerina-codes committed:
Prompts for player two

Katerina-codes committed:
Prompts for a game mode

- Corrected player types so they use greek letters.

Katerina-codes committed:
Prompts for move

Katerina-codes committed:
Annouces draw

Katerina-codes committed:
Announce winner

Katerina-codes committed:
Prompts user for their language choice

Katerina-codes committed:
Added displayComputerMoveWasCalled in GameTest

- Corrected FakeCommandLineUI boolean for displayComputerMoveWasCalled.
- Extracted getting game mode logic and passed it to newGame method.
- Refactored newGame method to remove duplication.

Katerina-codes committed:
Extracted language logic to method & updated test

Katerina-codes committed:
Added getLanguageSelection method

- Renamed askUserForLanguageType method to askUserForLanguageSelection.

Katerina-codes committed:
Get language from user selection

Katerina-codes committed:
Added option to select Greek Language

Katerina-codes committed:
Added Greek class to Game constructor

- Removed English from CommandLineUI constructor.
- Added language as a parameter for user interection methods.
- Rename English class variable 'english' to 'englishLanguage'.

Katerina-codes committed:
Added method to convert Greek moves to English

Katerina-codes committed:
Converted greek move for paper to english

- Corrected test name.

Katerina-codes committed:
Converts greek move for scissors to english

- Refactored code to remove If Statement. Moved logic to hashmap.

Katerina-codes committed:
Added method to convert winning move to greek

Katerina-codes committed:
Bug fix to convert 'draw' to Greek

- Added a translation for 'draw'.

Katerina-codes committed:
Grouped conversion logic together

Katerina-codes committed:
Refactored tests to remove duplication

jsuchy commented on src/main/game/Game.java:
```diff
@@ -1,39 +1,95 @@
     private UI inputOutput;
 
-    public Game(UI inputOutput, Rules rules) {
+    public Game(UI inputOutput, Rules rules, English englishLanguage, Greek greekLanguage) {
```
Can you think of a way to minimise the number of dependencies (and arguments) here? Do we ever need both english and greek?

jsuchy commented on src/main/game/Game.java:
```diff
@@ -1,39 +1,95 @@
+    public Game(UI inputOutput, Rules rules, English englishLanguage, Greek greekLanguage) {
         this.inputOutput = inputOutput;
         this.rules = rules;
+        this.english = englishLanguage;
```
Consistent naming is important, english and greekLanguage aren't consistent.

jsuchy commented on src/main/game/Game.java:
```diff
@@ -1,39 +1,95 @@
-            inputOutput.displayComputerMove(playerTwoMove.getMove());
-            findWinner(playerOneMove, playerTwoMove);
+            Moves computerMove = ComputerPlayer.playRandomMove();
+            if (language.equals(greekLanguage)) {
```
This is an explicit type-check in disguise. Can you make this polymorphic behaviour instead of doing this type of check?

jsuchy commented on src/main/game/Game.java:
```diff
@@ -1,39 +1,95 @@
+        inputOutput.askForMove(language, playerNumber);
         String playerMove = inputOutput.getMoveFromUser();
-        return CommandLineUI.stringToEnum(playerMove);
+        if (language.equals(greekLanguage)) {
```
Same thing here.

jsuchy commented on src/main/game/Game.java:
```diff
@@ -1,39 +1,95 @@
-        return CommandLineUI.stringToEnum(playerMove);
+        if (language.equals(greekLanguage)) {
+            playerMove = convertToEnglish(playerMove);
+            return CommandLineUI.stringToEnum(playerMove);
```
There's some duplication here with the return statement.

jsuchy commented on src/main/game/Game.java:
```diff
@@ -1,39 +1,95 @@
+    public void findWinner(Moves playerOneMove, Moves playerTwoMove, Language language) {
         String winningMove = rules.scoreGame(playerOneMove, playerTwoMove);
-        inputOutput.announceWinner(winningMove);
+        if (language.equals(greekLanguage)) {
```
Type check again

jsuchy commented on src/test/game/CommandLineUITest.java:
```diff
@@ -17,17 +18,35 @@
+    public void askUserForLanguageSelection() {
+        ui.askUserForLanguageSelection();
+
+        assertTrue(output.toString().contains("Select your language:\nEnter '1' for English\nΕισαγάγετε '2' για τα Ελληνικά"));
```
If the implementation of English changes, these tests will break. Perhaps you could implement your own TestLanguage to remove that dependency and to isolate these tests from English.

jsuchy commented on src/test/game/CommandLineUITest.java:
```diff
@@ -17,17 +18,35 @@
+
+       CommandLineUI ui = new CommandLineUI(new PrintStream(output), input);
+
+        assertEquals("1", ui.getLanguageSelection());
```
Make sure you leverage your IDE's ability to reformat your code to avoid stray spaces like this one.

Katerina-codes committed:
Moved logic to return correct language class

- All logic related to returning the correct language is now in
getLanguageSelection method.

Katerina-codes committed:
Made variable naming consistent for English class
